### PR TITLE
Allow overlapping boundaries in boundary conditions

### LINKED ISSUE #1205: Boundary dofs common to two boundary conditions get the wrong value
I have a domain that is a cube and a MeshTags object that marks triangles on different faces to different values. I apply Dirichlet boundary conditions to the various faces like so 

```
        auto dofs = fem::locate_dofs_topological(  // Find dofs
                {*V},                              // In this space (P1)
                2,                                 // Triangles
                meshTags.find(subdomainIndex));    // These triangles

        bc.push_back(
                std::make_shared<const fem::DirichletBC<PetscScalar>>(fun,
                                                                      dofs));
```
However, nodes on the edges between two faces end up having the wrong boundary condition applied to them.
I originally thought I understood why this happened, namely that the nodes were lifted twice, but I am now second-guessing that.

I was able to workaround the issue by manually removing duplicates from the output of `fem::locate_dofs_topological`, in which case I get the correct result.

### PR DESCRIPTION
**Pull request by massimiliano-leoni**
Closes #1205 

Adding a wrapper to be optionally used in `fem::add_diagonal` when boundary conditions have overlapping boundaries.

Suggestions of alternative implementations are welcome.

### CODE REVIEW

**Comment by massimiliano-leoni** [comment]
Alternatively, we could add a `fem::insert_diagonal` [or `fem::set_diagonal`?] function and then perhaps bind `fem::add_diagonal` to always use `fem::add_fn` and `fem::insert_diagonal` to use `fem::set_fun`.

**Comment by garth-wells** [comment]
Thanks for looking at this. A couple of questions:
- Does using `INSERT_VALUES` mean that we need to flush the matrix after assembler and before calling this function?
- The RHS still gets modified more than once with overlapping boundary conditions, I presume, with the boundary conditions summed? Is this desirable behaviour?

**Comment by massimiliano-leoni** [comment]
You are welcome!

- Yes, the matrix needs to be flushed whenever switching between `ADD_VALUES` and `INSERT_VALUES` modes. PETSc issues an error message if this is not done.
- The RHS is actually fine, because all the dofs are selected in advance and duplicates are removed [cfr. `assemble_vector_impl.h:871`]. If I understand correctly, the last boundary value in the list is the one that applies.

**Comment by michalhabera** [comment]
I agree this (allow to specify insert mode) is very useful, not only in the context of BCs. I added it some time ago, but the implementation using lambda functions is superior.

I only find the naming a bit confusing. I'd suggest to follow the PETSc naming, i.e. InsertMode could be either "insert" or "add". So if you make `add_fn` generic, it should be named something along`PETScMatrix::set_fn(Mat A, const InsertMode mode)` and specialised as `PETScMatrix::set_fn_add(Mat A)` for InsertMode = ADD and `PETScMatrix::set_fn_insert(Mat A)` for INSERT mode.

We'd need a python test for this as well, including pybind11 wrappers then.

**Comment by massimiliano-leoni** [comment]
@michalhabera thanks for the input! I don't mind following the naming convention you suggest, I can apply it if nobody objects.

On the Python tests and wrappers, I can look into that. Any pointers to where to start?

**Comment by garth-wells** [comment]
> * The RHS is actually fine, because all the dofs are selected in advance and duplicates are removed [cfr. `assemble_vector_impl.h:871`]. If I understand correctly, the last boundary value in the list is the one that applies.

Could you add a test for this? We need to be sure that it keeps working.

**Comment by garth-wells** [comment]
> I only find the naming a bit confusing. I'd suggest to follow the PETSc naming, i.e. InsertMode could be either "insert" or "add". So if you make `add_fn` generic, it should be named something along`PETScMatrix::set_fn(Mat A, const InsertMode mode)` and specialised as `PETScMatrix::set_fn_add(Mat A)` for InsertMode = ADD and `PETScMatrix::set_fn_insert(Mat A)` for INSERT mode.

I don't think we need more functions - the current naming follows the PETSc nomenclature.

**Comment by michalhabera** [comment]
> > I only find the naming a bit confusing. I'd suggest to follow the PETSc naming, i.e. InsertMode could be either "insert" or "add". So if you make `add_fn` generic, it should be named something along`PETScMatrix::set_fn(Mat A, const InsertMode mode)` and specialised as `PETScMatrix::set_fn_add(Mat A)` for InsertMode = ADD and `PETScMatrix::set_fn_insert(Mat A)` for INSERT mode.
> 
> I don't think we need more functions - the current naming follows the PETSc nomenclature.

The current naming (in master) is. But if the PR makes `add_fn` with optional `InsertMode`, it should not be named "add" anymore.

**Comment by garth-wells** [comment]
> > > I only find the naming a bit confusing. I'd suggest to follow the PETSc naming, i.e. InsertMode could be either "insert" or "add". So if you make `add_fn` generic, it should be named something along`PETScMatrix::set_fn(Mat A, const InsertMode mode)` and specialised as `PETScMatrix::set_fn_add(Mat A)` for InsertMode = ADD and `PETScMatrix::set_fn_insert(Mat A)` for INSERT mode.
> > 
> > 
> > I don't think we need more functions - the current naming follows the PETSc nomenclature.
> 
> The current naming (in master) does. But if the PR makes `add_fn` with optional `InsertMode`, it should not be named "add" anymore.

Ok, let's go with `PETScMatrix::set_fn` since this mirrors `MatSetValues` from PETSc which does both insert and add.

**Comment by massimiliano-leoni** [comment]
I updated the function names and I edited the Python wrappers. I added a function `insert_diagonal` that calls the new function that uses `INSERT_VALUES`.

File: python/test/unit/fem/test_bcs.py [hunk 1]
```diff
@@ -43,3 +45,47 @@ def test_locate_dofs_geometrical():
         # Check correct dof returned in V
         coords_V = V.tabulate_dof_coordinates()
         assert np.isclose(coords_V[dofs[0][1]], [0, 0, 0]).all()
+
+
+def test_overlapping_bcs():
```
Comment (garth-wells) [inline python/test/unit/fem/test_bcs.py:50]: Should this test also perform a solve and check that the solution matches the boundary condition?

**Review by garth-wells — CHANGES_REQUESTED** [review]

File: python/dolfinx/wrappers/fem.cpp [hunk 2]
```diff
@@ -376,9 +376,17 @@ void fem(py::module& m)
            const std::vector<std::shared_ptr<
                const dolfinx::fem::DirichletBC<PetscScalar>>>& bcs,
            PetscScalar diagonal) {
-          dolfinx::fem::add_diagonal(dolfinx::la::PETScMatrix::add_fn(A), V,
+          dolfinx::fem::add_diagonal(dolfinx::la::PETScMatrix::set_fn_add(A), V,
                                      bcs, diagonal);
         });
+  m.def("insert_diagonal",
```
[the remark this answers is not in the document]
Comment (massimiliano-leoni) [reply python/dolfinx/wrappers/fem.cpp:382]: It's not evident to me how to do this; `add_diagonal` has its own wrapper.

File: python/dolfinx/wrappers/la.cpp [hunk 3]
```diff
@@ -143,5 +143,13 @@ void la(py::module& m)
   //         },
   //         py::return_value_policy::take_ownership,
   //         "Create a PETSc MatNullSpace.");
+  m.def(
```
[the remark this answers is not in the document]
Comment (massimiliano-leoni) [reply python/dolfinx/wrappers/la.cpp:146]: Done.

File: cpp/dolfinx/la/PETScMatrix.h [hunk 4]
```diff
@@ -38,19 +38,62 @@ MatNullSpace create_petsc_nullspace(MPI_Comm comm,
...
-  /// Return a function with an interface for adding values to the
-  /// matrix A (calls MatSetValuesLocal)
+  /// Return a function with an interface for adding or inserting values
+  /// into the matrix A (calls MatSetValuesLocal)
   /// @param[in] A The matrix to set values in
+  /// @param[in] mode The PETSc insert mode (ADD_VALUES, INSERT_VALUES, ...)
   static std::function<int(std::int32_t, const std::int32_t*, std::int32_t,
                            const std::int32_t*, const PetscScalar*)>
-  add_fn(Mat A);
+  set_fn(Mat A, const InsertMode mode);
```
[the remark this answers is not in the document]
Comment (massimiliano-leoni) [reply cpp/dolfinx/la/PETScMatrix.h:47]: I thought you said to use that name to mirror `MatSetValues`. Do you want `insert_fn_add` and `insert_fn_insert` then?

File: python/test/unit/fem/test_bcs.py (hunk 1 above)
Comment (massimiliano-leoni) [reply python/test/unit/fem/test_bcs.py:50]: The solve shouldn't affect the boundary condition if the value in the rhs is correct and the corresponding diagonal entry in the lhs is 1.

File: cpp/dolfinx/la/PETScMatrix.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (michalhabera) [reply cpp/dolfinx/la/PETScMatrix.h:47]: We agreed `set_fn` is consistent with `MatSetValues`. Maybe just some confusion.

[the remark this answers is not in the document]
Comment (garth-wells) [reply cpp/dolfinx/la/PETScMatrix.h:47]: Sorry, top of my head I thought it was `MatInsertValues`, but the PETSc function is  `MatSetValues` so we want `set_fn`.

**Comment by massimiliano-leoni** [comment]
Ok, I think it's essentially ready, except for one thing: in `python/dolfinx/fem/assemble.py`, line 228 I implement the new functionality into `assemble_matrix`, where I first assemble/flush and then call the new function which inserts values instead of adding them. All good there.
On line 316, I try to do the same for the block assembly but I keep getting a PETSc error `RuntimeError: Failed to successfully call PETSc function 'MatSetValuesLocal'. PETSc error code is: 73, Object is in wrong state`.
Can anyone point out what I am doing wrong? I tried different combinations of assembly routines, on `A` and `Asub`, but to no avail.

This behaviour is currently disabled in the branch, but the lines are there, just commented out. Since it is disabled, when the CI runs the demos it shows, for Navier-Stokes, that the solution for the regular assembly [which uses the new imposition of the boundary conditions] and the block assembly [which uses the old one] are actually different.
